src/choice_map.hpp

```diff
@@ -32,123 +31,24 @@ class ChoiceMap {
   //   - Edge goes to leaf (NoId for left and right child).
   // - Edges span every leaf and root node.
-  bool SelectionIsValid(const bool is_quiet = true) const {
-    size_t edge_max_id = dag_.EdgeIdxRange().second;
-    for (size_t edge_idx = 0; edge_idx < edge_choice_vector_.size(); edge_idx++) {
-      const auto &edge_choice = edge_choice_vector_[edge_idx];
-      // If edge id is outside valid range.
-      if ((edge_choice.parent_edge_id > edge_max_id) ||
-          (edge_choice.sister_edge_id > edge_max_id)) {
-        // If they are not NoId, then it is an invalid edge_id.
-        if ((edge_choice.parent_edge_id != NoId) ||
-            (edge_choice.sister_edge_id != NoId)) {
-          if (!is_quiet) {
-            std::cerr << "Parent or Sister has invalid edge id." << std::endl;
-          }
-          return false;
-        }
-        // NoId is valid only if edge goes to a root.
-        if (!dag_.IsEdgeRoot(edge_idx)) {
-          if (!is_quiet) {
-            std::cerr << "Parent or Sister has NoId when edge is not a root."
-                      << std::endl;
-          }
-          return false;
-        }
-      }
-      for (const auto &child_edge_id :
-           {edge_choice.left_child_edge_id, edge_choice.right_child_edge_id}) {
-        // If edge id is outside valid range.
-        if (child_edge_id > edge_max_id) {
-          // If they are not NoId, then it is an invalid edge_id.
-          if (child_edge_id != NoId) {
-            if (!is_quiet) {
-              std::cerr << "Child has invalid edge id." << std::endl;
-            }
-            return false;
-          }
-          // NoId is valid only if edge goes to a leaf.
-          if (!dag_.IsEdgeLeaf(edge_idx)) {
-            if (!is_quiet) {
-              std::cerr << "Child has NoId when edge is not a leaf." << std::endl;
-            }
-            return false;
-          }
-        }
-      }
-    }
-    return true;
-  }
+  bool SelectionIsValid(const bool is_quiet = true) const;
 
   // ** TreeMask
-  // A TreeMask is an unordered vector of edge Ids, which represent a tree contained in
+  // A TreeMask is a set of edge Ids, which represent a tree contained in
   // the DAG, from the selected subset of DAG edges.
-  using TreeMask = std::vector<size_t>;
-
-  // Convert stack to a vector.
-  template <typename T>
-  static std::set<T> StackToVector(std::stack<T> &stack) {
-    T *end = &stack.top() + 1;
-    T *begin = end - stack.size();
-    std::vector<T> stack_contents(begin, end);
-    return stack_contents;
-  }
-
-  // Push Id to Stack if it is not NoId.
+  using TreeMask = std::set<size_t>;
+  enum class AdjacentNode { Parent, LeftChild, RightChild };
   template <typename T>
-  static void StackPushIfValidId(std::stack<T> &stack, size_t id) {
-    if (id != NoId) {
-      stack.push(id);
-    }
-  }
+  using AdjacentNodeArray = EnumArray<AdjacentNode, 3, T>;
+  using ExpandedTreeMask = std::map<size_t, AdjacentNodeArray<size_t>>;
```

Comment:
I'm just noting that if we had a NodeId type this typedef would be a bunch clearer!

Reply:
Yes, we need to make `NodeId` and `EdgeId` strongly typed across Bito.  We already have an issue raised for that.  I'll work on that this week!